File: tests/test_module.cpp
/**
 * Test: test_module
 * Tests LLVM Module creation and properties
 *
 * LLVM-C APIs covered:
 * - LLVMModuleCreateWithNameInContext()
 * - LLVMGetModuleIdentifier() / LLVMSetModuleIdentifier()
 * - LLVMGetSourceFileName() / LLVMSetSourceFileName()
 * - LLVMGetDataLayoutStr() / LLVMSetDataLayout()
 * - LLVMGetTarget() / LLVMSetTarget()
 * - LLVMPrintModuleToString()
 * - LLVMCloneModule()
 * - LLVMDisposeModule()
 */

#include <llvm-c/Core.h>
#include <llvm-c/Analysis.h>
#include <cstdio>
#include <cstring>
#include <string>

int main() {
    LLVMContextRef ctx = LLVMContextCreate();

    // Create module
    LLVMModuleRef mod = LLVMModuleCreateWithNameInContext("test_module", ctx);

    // Get initial module identifier (copy to string before modifying)
    size_t id_len;
    const char *id_ptr = LLVMGetModuleIdentifier(mod, &id_len);
    std::string initial_id(id_ptr, id_len);

    // Set new module identifier
    LLVMSetModuleIdentifier(mod, "renamed_module", strlen("renamed_module"));
    id_ptr = LLVMGetModuleIdentifier(mod, &id_len);
    std::string new_id(id_ptr, id_len);

    // Get/set source filename (copy to string before modifying)
    size_t src_len;
    const char *src_ptr = LLVMGetSourceFileName(mod, &src_len);
    std::string initial_src(src_ptr, src_len);
    LLVMSetSourceFileName(mod, "test_source.c", strlen("test_source.c"));
    src_ptr = LLVMGetSourceFileName(mod, &src_len);
    std::string new_src(src_ptr, src_len);

    // Get/set data layout (copy to string before modifying)
    std::string initial_layout = LLVMGetDataLayoutStr(mod);
    LLVMSetDataLayout(mod, "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128");
    std::string new_layout = LLVMGetDataLayoutStr(mod);

    // Get/set target triple (copy to string before modifying)
    std::string initial_target = LLVMGetTarget(mod);
    LLVMSetTarget(mod, "x86_64-unknown-linux-gnu");
    std::string new_target = LLVMGetTarget(mod);

    // Clone module
    LLVMModuleRef cloned = LLVMCloneModule(mod);
    id_ptr = LLVMGetModuleIdentifier(cloned, &id_len);
    std::string cloned_id(id_ptr, id_len);

    // Verify module
    char *error = nullptr;
    if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
        fprintf(stderr, "; Verification failed: %s\n", error);
        LLVMDisposeMessage(error);
        LLVMDisposeModule(cloned);
        LLVMDisposeModule(mod);
        LLVMContextDispose(ctx);
        return 1;
    }
    LLVMDisposeMessage(error);

    // Print diagnostic comments
    printf("; Test: test_module\n");
    printf("; Initial module ID: %s\n", initial_id.c_str());
    printf("; New module ID: %s\n", new_id.c_str());
    printf("; Initial source filename: %s\n", initial_src.c_str());
    printf("; New source filename: %s\n", new_src.c_str());
    printf("; Initial data layout: %s\n", initial_layout.empty() ? "(empty)" : initial_layout.c_str());
    printf("; New data layout: %s\n", new_layout.c_str());
    printf("; Initial target: %s\n", initial_target.empty() ? "(empty)" : initial_target.c_str());
    printf("; New target: %s\n", new_target.c_str());
    printf("; Cloned module ID: %s\n", cloned_id.c_str());
    printf("\n");

    // Print module IR
    char *ir = LLVMPrintModuleToString(mod);
    printf("%s", ir);
    LLVMDisposeMessage(ir);

    // Cleanup
    LLVMDisposeModule(cloned);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);

    return 0;
}

File: test_module.py
#!/usr/bin/env -S uv run
"""
Test: test_module
Tests LLVM Module creation and properties via Python bindings.

This is the Python equivalent of tests/test_module.cpp.
Output should match the C++ golden master test.

LLVM APIs covered (via Python bindings):
- Module creation (via context manager)
- name property (get/set)
- source_filename property (get/set)
- data_layout property (get/set)
- target_triple property (get/set)
- to_string()
- clone()
- verify()
"""

import sys
import llvm


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_module") as mod:
            # Get initial module identifier
            initial_id = mod.name

            # Set new module identifier
            mod.name = "renamed_module"
            new_id = mod.name

            # Get/set source filename
            initial_src = mod.source_filename
            mod.source_filename = "test_source.c"
            new_src = mod.source_filename

            # Get/set data layout
            initial_layout = mod.data_layout
            mod.data_layout = (
                "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
            )
            new_layout = mod.data_layout

            # Get/set target triple
            initial_target = mod.target_triple
            mod.target_triple = "x86_64-unknown-linux-gnu"
            new_target = mod.target_triple

            # Clone module - returns a ModuleManager, must use 'with' or .dispose()
            with mod.clone() as cloned:
                cloned_id = cloned.name

            # Verify module
            if not mod.verify():
                print(
                    f"; Verification failed: {mod.get_verification_error()}",
                    file=sys.stderr,
                )
                return 1

            # Print diagnostic comments
            print("; Test: test_module")
            print(f"; Initial module ID: {initial_id}")
            print(f"; New module ID: {new_id}")
            print(f"; Initial source filename: {initial_src}")
            print(f"; New source filename: {new_src}")
            print(
                f"; Initial data layout: {initial_layout if initial_layout else '(empty)'}"
            )
            print(f"; New data layout: {new_layout}")
            print(
                f"; Initial target: {initial_target if initial_target else '(empty)'}"
            )
            print(f"; New target: {new_target}")
            print(f"; Cloned module ID: {cloned_id}")
            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    sys.exit(main())
